lib-meshrabiya-vpn/src/main/java/com/meshrabiya/lib_vpn/MeshrabiyaVpnService.kt

```diff
@@ -0,0 +1,306 @@
+    }
+
+    /**
+     * Builds an IPv4 address from the provided buffer.
+     * @param buffer the buffer containing the packet data.
+     * @return the constructed IPv4 address.
+     */
+    private fun buildIPv4Address(buffer: ByteBuffer): InetAddress {
+        return InetAddress.getByAddress(ByteArray(4).also { buffer.get(it, 0, 4) })
+    }
+
+    /**
+     * Parses the IPv4 header to extract the destination port and protocol name.
+     * @param buffer the buffer containing the IPv4 packet data.
+     * @return a pair of the destination port and the protocol name.
+     */
+    private fun parseIPv4Header(buffer: ByteBuffer): Pair<Int, String> {
+        try {
+            val protocol = buffer.get(IPV4_PROTOCOL_OFFSET).toInt()
+            val isTcp = protocol == IPPROTO_TCP
+            val isUdp = protocol == IPPROTO_UDP
+
+            val ipHeaderLength = (buffer.get(IPV4_HEADER_LENGTH_OFFSET).toInt() and 0x0F) * 4
+            val destinationPort = if (isTcp || isUdp) {
+                ((buffer.get(ipHeaderLength + 2).toInt() and 0xFF) shl 8) or
+                        (buffer.get(ipHeaderLength + 3).toInt() and 0xFF)
+            } else {
+                -1
+            }
+
+            val protocolName = when {
+                isTcp -> "TCP"
+                isUdp -> "UDP"
+                else -> "Unknown"
+            }
+
+            return Pair(destinationPort, protocolName)
+        } catch (e: Exception) {
+            logMessage("Error parsing IPv4 header: ${e.message}")
+            return Pair(-1, "Error")
+        }
+    }
+
+    /**
+     * Handles an IPv6 packet by extracting the destination IP and port, and processing it based on the network type.
+     * @param buffer the buffer containing the IPv6 packet data.
+     * @param length the length of the packet data.
+     */
+    private fun handleIPv6Packet(buffer: ByteBuffer, length: Int) {
+        val destinationIp = buildIPv6Address(buffer)
+        val (destinationPort, protocolName) = parseIPv6Header(buffer)
+
+        logMessage("IPv6 Packet - Destination: [${destinationIp.hostAddress}]:$destinationPort, Protocol: $protocolName")
+
+        if (isVirtualNetwork(destinationIp)) {
+            handleVirtualNetworkPacket(buffer, length)
+        } else {
+            writeToVpnInterface(buffer)
+        }
+    }
+
+    /**
+     * Builds an IPv6 address from the provided buffer.
+     * @param buffer the buffer containing the packet data.
+     * @return the constructed IPv6 address.
+     */
+    private fun buildIPv6Address(buffer: ByteBuffer): InetAddress {
+        return InetAddress.getByAddress(ByteArray(16).also { buffer.get(it, 0, 16) })
+    }
+
+    /**
+     * Parses the IPv6 header to extract the destination port and protocol name.
+     * @param buffer the buffer containing the IPv6 packet data.
+     * @return a pair of the destination port and the protocol name.
+     */
+    private fun parseIPv6Header(buffer: ByteBuffer): Pair<Int, String> {
```

Comment:
parseIpHeader functions should return a data class (e.g. IpHeader) - that should contain destination address (InetAddress), destination port, origin address (InetAddress), origin port, and protocol.

Reply:
Note we need to know the origin port because there could be multiple streams going to the same destination. For Meshrabiya to match things up correctly, we will need to know the origin.